### ISSUE TITLE
Can port forwarding be done with dadb?

### ISSUE THREAD

**Issue by mrubinsquare** [mobile-dev-inc/dadb#1]
Hi, there. I've been enjoying playing with this library. Thank you so much for it!

I have a need to set up [port forwarding](https://developer.android.com/studio/command-line/adb?hl=it#forwardports), as in `adb -s emulator-5554 forward tcp:6100 tcp:7100`. I suspect this requires the adb server itself, but I'm still green, figuring out how the protocols work. I've tried my own hand at adding a reboot command, via a simple `dadb.open("reboot:")`, and that worked. So I toyed with something like that I thought might work for forwarding (e.g. `dadb.open("host-serial:5554:forward:tcp:10001;tcp:10001)`) based on looking at the `ddmlib` source [for forwarding](https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-main:ddmlib/src/main/java/com/android/ddmlib/AdbHelper.java;l=901?q=createForward), but I barely know what I'm doing. Obviously, that didn't work for me. Any advice?

Thanks again for the great library and any help.

**Comment by Leland-Takamine** [comment]
Hey @mrubinsquare - great question! This particular command is one that the ADB Server handles. dadb circumvents the ADB Server which is why `dadb.open("...:forward:...)` won't work.

However, there are ways to implement what you need:

`val stream = dadb.open("tcp:10001")` will open an `AdbStream` that you can use to communicate with port `10001` on the device. Depending on what you need, you have a couple options from there:

* If you only need your program to be able to access port `10001` on the device, you can use `AdbStream.sink / source` to send data to / from the device port.
* If you need other processes to be able to connect to port `10001` on the device, then you can kick off a thread in your program to handle the port forwarding manually.

I'd be interested to know more about your specific needs to understand whether it make sense to build a higher-level abstraction to better support this use case.

**Comment by mrubinsquare** [comment]
Thanks so much for the reply! It makes sense that this requires the adb server. In case other folks come here, I dug a further to validate that by looking at the adb server source code, starting [here](https://android.googlesource.com/platform/system/adb/+/refs/heads/master/adb.cpp#959).

So my use case is not too dissimilar from the [Interconnecting emulator instances](https://developer.android.com/studio/run/emulator-networking) example in the Android developer docs: we have an app on one device that wants to connect to another app running on another device over tcp at a well known port; when we run each in their own emulator (on the same host machine), we have devs set up adb port forwarding to allow the second, server emulator be reachable at that port. After seeing your library's release, it occurred to me that maybe I could compile an adb client into the server app (development build only) that automatically (on emulators only) does the port forwarding when the server app starts up, making things a bit more ergononic/automatic/foolproof for our devs. I had success with using [adam](https://github.com/Malinskiy/adam) for this (configuring it to hit 10.0.2.2, the emulator's alias the host machine's loopback interface, to do the forwarding). But for a variety of reasons, I wanted to see if I could get it working with dadb instead. 

Now that I have confirmation that adb port forwarding does require the adb server, my use case does seem at cross purposes with the whole point of your nifty library, but I really am still learning here, so if there's a way with dadb, I'm very interested.

**Comment by Leland-Takamine** [comment]
Ok so after digging a bit - we could add support for this in dadb. But after thinking through your use case, it seems like the [QEMU pipe device](https://android.googlesource.com/platform/external/qemu/+/master/docs/ANDROID-QEMU-PIPE.TXT) might be a more elegant approach.

What this device allows your app to to do is listen on a tcp port on the host by opening up and reading/writing to the `/dev/qemu_pipe` device. Your server app can use this to listen directly on a well-known tcp port on the host without any port forwarding.

Now the issue I'm running into is that I'm not able to get this working properly:

```kotlin
val file = File("/dev/goldfish_pipe")
file.outputStream().writer().use { out ->
    out.write("tcp:8000\u0000")
    out.flush()
    Thread.sleep(10000)
}
```

The code above is failing for me with `EINVAL (Invalid argument)`.

It might be worthwhile for you to read through [these docs](https://android.googlesource.com/platform/external/qemu/+/master/docs/ANDROID-QEMU-PIPE.TXT) and try your hand at it. Maybe you'll figure something out that I missed. If you're able to get this working, it'll drastically simplify your setup.

If you try the above and determine that port forwarding is still the right approach we can look into implementing this into dadb. Please let me know either way!

**Comment by mrubinsquare** [comment]
Oh, I completely missed your update. Thank you so much for the work and the suggestion. I'll take a look at this.

**Comment by mrubinsquare** [comment]
First, thanks again for pointing me at this. I got a lot of value even just hearing about QEMU pipe device and (hopefully) deciphering its purpose and learning a bit more about QEMU and the like.

Okay, I'm really not terribly expert on any of the related topics, but I think I get the idea here. Please correct me if I'm wrong. We currently implement our TCP communication logic with a Java `SocketConnection`. So I think if we went this way, we'd be replacing the `read`s and `write`s there with `read`s and `write`s to the `FileDescriptor` from the QEMU pipe device, assuming we can get that to work from within an Android app running on the emulator. I think this ends up being a fair amount of refactoring and a loss of abstractions which we'd have to reimplement? We have working layered networking code that boils down to this TCP connection I'd like not to fiddle with. Although the performance will be better with the QEMU pipe device, I think that's lower level than I'd like to be. We're not so interested in the performance improvement on the emulators in our context, but the ease with which the client and server running on the two different emulators can connect. Port forwarding is closer to that abstraction level, leaving the rest of the implementation (that services several other contexts) untouched. But really, I'm dabbling in areas I don't claim to be very informed on (networking, sockets, tcp, C).

I can make do with using `adam`, I think, to do the port forwarding from the app pretty straightforwardly, though I still would love to use the lighter `dadb`, but of course, it's at your discretion if, when, and how you implement something closer to that. No pressure or complaints from me. I appreciate your time and help!

**Comment by Leland-Takamine** [comment]
@mrubinsquare Yes your understanding is correct. Your underlying server I/O logic would need to be updated to read/write to the qemu device FileDescriptor. It makes sense that it's hard to justify this sort of refactoring. I'll bring this up with the team to see if it's something we can tackle in the near term.

**Comment by mrubinsquare** [comment]
Thank you!

**Comment by axelniklasson** [comment]
Hello @mrubinsquare, just wanted to follow up here and let you know that we just published [v0.0.10](https://mvnrepository.com/artifact/dev.mobile/dadb/0.0.10) of `dadb` that includes TCP forwarding support. Thanks again for the suggestion, it is much appreciated!

**Comment by mrubinsquare** [comment]
That's really exciting and thanks for the update here, the work itself, and taking the suggestion so seriously! I'll try to find time early next week to give this a go.
